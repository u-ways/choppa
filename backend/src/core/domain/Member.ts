export default interface Member {
  readonly id: string;
  readonly name: string;
  // readonly role: Chapter;
}
