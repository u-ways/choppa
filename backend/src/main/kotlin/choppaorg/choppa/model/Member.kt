package choppaorg.choppa.model

import javax.persistence.*

@Entity
class Member(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Int?,

    var name: String,

    @OneToOne
    @JoinColumn(name = "chapter_id")
    var chapter: Chapter?
) {
    constructor(name: String, chapter: Chapter): this(null, name, chapter)
}