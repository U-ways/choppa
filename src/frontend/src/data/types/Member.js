export default class Member {
  constructor(id, name, chapter) {
    this._id = id;
    this._name = name;
    this._chapter = chapter;
  }

  get id() {
    return this._id;
  }

  get name() {
    return this._name;
  }

  set name(newName) {
    this._name = newName;
  }

  get chapter() {
    return this._chapter;
  }
}
