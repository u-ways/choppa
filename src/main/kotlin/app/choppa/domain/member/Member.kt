package app.choppa.domain.member

import app.choppa.domain.base.BaseModel
import app.choppa.domain.chapter.Chapter
import app.choppa.domain.chapter.Chapter.Companion.UNASSIGNED_ROLE
import app.choppa.domain.squad.Squad
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.hibernate.annotations.GenericGenerator
import java.util.*
import java.util.UUID.randomUUID
import javax.persistence.*

@Entity
@Table(name = "member")
@JsonSerialize(using = MemberSerializer::class)
@JsonDeserialize(using = MemberDeserializer::class)
data class Member(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "ME-$id".substring(0, 15),

    @ManyToOne
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter = UNASSIGNED_ROLE,

    @Column(name = "active", columnDefinition = "BOOLEAN")
    val active: Boolean = true,

    @ManyToMany(mappedBy = "members")
    val squads: MutableList<Squad> = mutableListOf(),
) : BaseModel {
    override fun toString() = "Member(id=$id, name=$name, chapter=${chapter.name})"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Member
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_MEMBERS get() = mutableListOf<Member>()
    }
}
