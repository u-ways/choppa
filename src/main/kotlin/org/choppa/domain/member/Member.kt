package org.choppa.domain.member

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.choppa.domain.base.BaseModel
import org.choppa.domain.chapter.Chapter
import org.choppa.domain.chapter.Chapter.Companion.UNASSIGNED_ROLE
import org.choppa.domain.history.History
import org.choppa.domain.squad.Squad
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "member")
@JsonSerialize(using = MemberSerializer::class)
@JsonDeserialize(using = MemberDeserializer::class)
data class Member(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "ME-$id".substring(0, 15),

    @ManyToOne
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter = UNASSIGNED_ROLE,

    @ManyToMany(mappedBy = "members")
    val squads: MutableList<Squad> = mutableListOf(),

    @OneToMany(mappedBy = "member")
    val history: MutableList<History> = mutableListOf()
) : BaseModel {
    override fun toString() = "Member(id=$id, name=$name, chapter=$chapter)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Member
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_MEMBERS get() = mutableListOf<Member>()
    }
}
