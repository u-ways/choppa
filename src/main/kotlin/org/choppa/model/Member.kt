package org.choppa.model

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import org.choppa.model.relations.IterationHistory
import org.choppa.model.relations.SquadCurrentMembers
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType.EAGER
import javax.persistence.FetchType.LAZY
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "member")
data class Member @JsonCreator constructor(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @JsonProperty("id")
    val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    @JsonProperty("name")
    val name: String,

    @ManyToOne(fetch = EAGER)
    @JsonProperty("chapter")
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter,

    @OneToMany(mappedBy = "member", fetch = EAGER)
    @JsonIgnore
    var squads: MutableList<SquadCurrentMembers> = mutableListOf(),

    @OneToMany(mappedBy = "member", fetch = LAZY)
    @JsonIgnore
    var iterations: MutableList<IterationHistory> = mutableListOf()
) {
    override fun toString() = "Member(id=$id, name=$name, chapter=$chapter)"
}
