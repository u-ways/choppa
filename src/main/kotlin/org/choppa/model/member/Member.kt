package org.choppa.model.member

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import org.choppa.model.chapter.Chapter
import org.choppa.model.history.History
import org.choppa.model.squad.Squad
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType.EAGER
import javax.persistence.FetchType.LAZY
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "member")
data class Member @JsonCreator constructor(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @JsonProperty("id")
    val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    @JsonProperty("name")
    val name: String,

    @ManyToOne(fetch = EAGER)
    @JsonProperty("chapter")
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter,

    @ManyToMany(mappedBy = "members", fetch = EAGER)
    @JsonIgnore
    val squads: MutableList<Squad> = mutableListOf(UNASSIGNED_SQUAD),

    @OneToMany(mappedBy = "member", fetch = LAZY)
    @JsonIgnore
    val iterations: MutableList<History> = mutableListOf()
) {
    override fun toString() = "Member(id=$id, name=$name, chapter=$chapter)"
}
