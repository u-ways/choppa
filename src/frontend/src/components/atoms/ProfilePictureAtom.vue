<template>
  <div class="profile-picture" :class="[ this.cssClasses ]">
    <canvas v-show="hasRenderedProfilePicture"
            width="100"
            height="100"
            ref="canvas"
            class="profile-picture__actual profile-picture__border"
    />
    <font-awesome-icon v-show="!hasRenderedProfilePicture"
                       icon="user-alt"
                       class="profile-picture__default-picture profile-picture__border"
    />
  </div>
</template>

<script>
import Member from "@/data/types/Member";
import { update } from "jdenticon";

export default {
  name: "ProfilePictureAtom",
  props: {
    member: {
      type: Member,
      required: true,
    },
    cssClasses: String,
  },
  data() {
    return {
      hasRenderedProfilePicture: false,
    };
  },
  mounted() {
    this.renderProfilePicture();
  },
  methods: {
    renderProfilePicture() {
      this.$nextTick(() => {
        update(this.$refs.canvas, `${this.member.id}${this.member.name}`);
        this.hasRenderedProfilePicture = true;
      });
    },
  },
};
</script>

<style scoped lang="scss">
@import "../../assets/scss/colours";
@import "../../assets/scss/typography";

.profile-picture {
  width: 100%;
  height: 100%;
  background: $white;
  border-radius: 50%;
  overflow: hidden;

  &__default-picture {
    width: 100%;
    height: 100%;
    color: hsla(0, 0, 0, 0.7);
  }

  &__actual {
    background: $white;
    width: 100%;
    height: 100%;
    overflow: hidden;;
  }

  &__border {
    border: 2px solid $white-border;
    border-radius: 50%;
  }
}
</style>
