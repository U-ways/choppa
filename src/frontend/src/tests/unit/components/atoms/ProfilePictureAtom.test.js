/* eslint-disable import/first */
jest.mock("jdenticon"); // eslint-disable-line no-undef

import * as jdenticon from "jdenticon";
import { describe, expect, it, shallowMount } from "@/tests/unit/test-imports";
import ProfilePictureAtom from "@/_old/components/atoms/ProfilePictureAtom";
import Member from "@/data/types/member";

const member = new Member(1, "Taylor");

describe("Profile Picture Atom test", () => {
  it("Render Profile Picture updates the canvas and sets the flag", async () => {
    const wrapper = shallowMount(ProfilePictureAtom, {
      propsData: {
        member,
        css: "my-css-class-test",
      },
      stubs: {
        "font-awesome-icon": true,
      },
    });

    await wrapper.vm.$nextTick();
    jdenticon.update.mockClear();
    await wrapper.setData({ hasRenderedProfilePicture: false });

    wrapper.vm.renderProfilePicture();
    await wrapper.vm.$nextTick();
    expect(jdenticon.update).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.$data.hasRenderedProfilePicture).toBeTruthy();
  });

  it("Renders correctly with fallback image", async () => {
    const wrapper = shallowMount(ProfilePictureAtom, {
      propsData: {
        member,
        css: "my-css-class-test",
      },
      stubs: {
        "font-awesome-icon": true,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Renders correctly with user's icon", async () => {
    const wrapper = shallowMount(ProfilePictureAtom, {
      propsData: {
        member,
        css: "my-css-class-test",
      },
      stubs: {
        "font-awesome-icon": true,
      },
    });

    await wrapper.setData({ hasRenderedProfilePicture: true });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
