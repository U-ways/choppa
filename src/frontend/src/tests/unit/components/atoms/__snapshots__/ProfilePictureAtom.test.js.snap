// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Picture Atom test Renders correctly with fallback image 1`] = `
"<div class=\\"profile-picture my-css-class-test\\"><canvas width=\\"100\\" height=\\"100\\" class=\\"profile-picture__actual profile-picture__border\\" style=\\"display: none;\\"></canvas>
  <font-awesome-icon-stub icon=\\"user-alt\\" class=\\"profile-picture__default-picture profile-picture__border\\"></font-awesome-icon-stub>
</div>"
`;

exports[`Profile Picture Atom test Renders correctly with user's icon 1`] = `
"<div class=\\"profile-picture my-css-class-test\\"><canvas width=\\"100\\" height=\\"100\\" class=\\"profile-picture__actual profile-picture__border\\" style=\\"\\"></canvas>
  <font-awesome-icon-stub icon=\\"user-alt\\" class=\\"profile-picture__default-picture profile-picture__border\\" style=\\"display: none;\\"></font-awesome-icon-stub>
</div>"
`;
